src/components/LocalFileUpload/custom-request-upload.ts

```diff
@@ -1,39 +1,113 @@
 import { CLEAR_SIMULARIUM_FILE } from "../../state/trajectory/constants";
 import { store } from "../..";
+
+// Typescript's File definition is missing this function
+//  which is part of the HTML standard on all browsers
+//  and needed below
+interface FileHTML extends File {
+    text(): Promise<string>;
+}
+
+let isLoading = false;
+
+/*
+A "custom request" for an Antd Upload component is a function that overrides
+the default POST request that happens when the user uploads files. The Antd
+Upload is a wrapper around the Upload component from react-component (rc-upload).
+- Antd docs: https://ant.design/components/upload/#API
+- rc-upload docs: https://github.com/react-component/upload#customrequest
+
+This custom request function takes in an array of all the files loaded by the
+user, finds and processes the .simularium file and any geometry files separately, 
+then tells the app to load the trajectory with its associated geometry files 
+into the viewer.
+*/
 export default (
-    { file, onSuccess, onError }: RcCustomRequestOptions,
+    { onSuccess, onError }: UploadRequestOption,
+    droppedFiles: File[],
     loadFunction: (simulariumFile: LocalSimFile) => void
 ) => {
+    if (isLoading === false) {
+        isLoading = true;
+    } else {
+        // If the user loads multiple files at once (.simularium file + geometry file(s)),
+        // this function is called multiple times, but we only need to process
+        // and load the trajectory once
+        return;
+    }
+
     // want the loading indicator to show without any lag time
     // as soon as user hits "Open" button, and not have to have this action called
     // multiple places in the code.
     store.dispatch({
         payload: { newFile: true },
         type: CLEAR_SIMULARIUM_FILE,
     });
-    file.text()
-        .then((text) => JSON.parse(text))
-        .then((data) => {
-            loadFunction({
-                lastModified: file.lastModified,
-                name: file.name,
-                data,
-            });
-        })
-        .then(() =>
-            onSuccess(
-                {
-                    name: file.name,
-                    status: "done",
-                    url: "",
-                },
-                file
-            )
-        )
-        .catch((error) => {
-            console.log(error);
-            onError(error);
-        });
+
+    const files: FileHTML[] = Array.from(droppedFiles) as FileHTML[];
+
+    Promise.all(files.map((file: FileHTML) => file.text())).then(
+        (parsedFiles: string[]) => {
+            try {
+                const simulariumFileIndex = findIndex(files, (file) =>
+                    file.name.includes(".simularium")
+                );
+                // TODO: handle when user doesn't load a .simularium file
+                // if (simulariumFileIndex === -1) {
+                //    throw new Error("Please upload a .simularium file.")
+                // }
```

Comment:
why is this code commented out? Is it not sufficient to throw Error?

Reply:
It's not sufficient -- I need to do something with it in the `catch` block in line 100